qiskit/transpiler/passes/utils/remove_final_measurements.py

```diff
@@ -19,12 +19,51 @@
 class RemoveFinalMeasurements(TransformationPass):
     """Remove final measurements and barriers at the end of a circuit.
 
-    This pass removes final barriers and final measurements, as well as the
-    ClassicalRegisters they are connected to if the ClassicalRegister
-    is unused. Measurements and barriers are considered final if they are
+    This pass removes final barriers and final measurements, as well as all
+    unused classical registers and bits they are connected to.
+    Measurements and barriers are considered final if they are
     followed by no other operations (aside from other measurements or barriers.)
+
+    Classical registers are removed iff they reference at least one bit
+    that has become unused by the circuit as a result of the operation, and all
+    of their other bits are also unused. Seperately, classical bits are removed
+    iff they have become unused by the circuit as a result of the operation,
+    or they appear in a removed classical register, but do not appear
+    in a classical register that will remain.
     """
 
+    def _calc_final_ops(self, dag):
+        final_op_types = {"measure", "barrier"}
+        final_ops = []
+
+        final_qubit_inputs = (next(dag.predecessors(dag.output_map[qubit])) for qubit in dag.qubits)
+        to_visit = list(final_qubit_inputs)
```

Comment:
Is there a reason to not just do?:

```suggestion
        to_visit = list(next(dag.predecessors(dag.output_map[qubit])) for qubit in dag.qubits)
```

Reply:
No real reason. I think a comment describing the role `to_visit` plays and what its initialized to would serve the original purpose even better. I'll add that as well.